runtime/lua/vscode/api.lua

```diff
@@ -418,4 +418,94 @@ do
   end
 end
 
+-----------------------
+------ utilities ------
+-----------------------
+
+---Perform operations in insert mode or switch to insert mode.
+---If in visual mode, this function will preserve the selection after
+---switching to insert mode.
+---
+---Example:
+---Make `editor.action.addSelectionToNextFindMatch` work correctly in any mode.
+---This is the behavior of the default VSCode shortcut Ctrl+d: Add Selection To Next Find Match.
+---
+---```lua
+---vim.keymap.set({ "n", "x", "i" }, "<C-d>", function()
+---  vscode.into_insert(function()
+---    vscode.action("editor.action.addSelectionToNextFindMatch")
+---  end)
+---end)
+---````
+---
+---@param callback function|nil Optional callback function to run after switching to insert mode.
+---                             If not provided, it will directly switch to insert mode synchronously.
+function M.into_insert(callback)
+  vim.validate({ callback = { callback, "f", true } })
+
+  local mode = api.nvim_get_mode().mode
+
+  --- Insert ---
+
+  if mode == "i" then
+    if callback then
+      callback()
+    end
+    return
+  end
+
+  --- Normal ---
+
+  if mode == "n" then
+    api.nvim_feedkeys("a", "n", false)
+    if callback then
+      callback()
+    else
+      vim.wait(100, function() end)
+    end
+    return
+  end
+
+  --- Visual ---
+
+  if mode:match("[vV\x16]") then
+    local A = fn.getpos("v")
+    local B = fn.getpos(".")
+    local start_pos = { A[2], A[3] - 1 }
+    local end_pos = { B[2], B[3] - 1 }
+
+    if start_pos[1] > end_pos[1] or (start_pos[1] == end_pos[1] and start_pos[2] > end_pos[2]) then
+      start_pos, end_pos = end_pos, start_pos
+    end
+
+    if mode == "V" then
+      start_pos = { start_pos[1], 0 }
+      end_pos = { end_pos[1], #fn.getline(end_pos[1]) }
+    end
+
+    local range = vim.lsp.util.make_given_range_params(start_pos, end_pos, 0, "utf-16").range
```

Comment:
Should probably split this up into util.visual_to_range

Reply:
It's not necessary for now. Although we currently have several modules that need to parse the selected range, their logic differs.

Reply:
Functions should usually have one function. This one has two: convert the selection to a range, and then send that range to vscode. It should be a util for readability, maintainability, and so it is available for future maintainers.

Reply:
- For multi-cursor, we've considered blockwise and treat linewise as a range line by line.
- For selection syncing, we've considered blockwise as well, but like charwise, we only consider the start and end positions of the selected region.
- In `into_insert`, we treat all as charwise (meaning we don't support blockwise, which isn't useful).

We can refactor in another PR, but only by separating functions. Otherwise, the function will become too complex.